### PR TITLE
DAOS-7174 tools: Add label support to daos util

### PR DESCRIPTION
**Pull request by mjmac**
With this commit, it is now possible to specify pool/container labels or
UUIDs or some combination thereof:
  $ daos cont query pool container
  $ daos cont query aa3667e4-63d8-49b3-9146-934c71700cc9 container
  $ daos cont query pool fc551f6c-88af-4a8d-8346-dcb40842fc96

Also lands some code cleanups and much-needed unit test coverage,
as well as a few bugfixes and review nits from the base PR.

Signed-off-by: Michael MacDonald <mjmac.macdonald@intel.com>

### CODE REVIEW

File: utils/node_local_test.py
Comment (ashleypittman) [inline utils/node_local_test.py:1952]: There is only one pool created by NLT (with the label NLT) and I'd deliberately left it so that the uuid was used in some instances and the label used in others.  I'm not sure if it's still the case after this PR but it would be good to check and keep that usage somewhere.

There's also currently a bug where pool is given by uuid but container by name so maybe it would be better to write a test specifically for this.

File: src/control/cmd/daos/util.h
Comment (ashleypittman) [inline src/control/cmd/daos/util.h:92]: D_STRNDUP() will log this anyway which is part of the benefit of using it.

File: src/control/cmd/daos/util.h
Comment (ashleypittman) [inline src/control/cmd/daos/util.h:80]: probably OK but there's a namespace conflict on "strlen" here, there's a chance it'll be interpreted as a function pointer rather than a int.  It should probably also be a size_t, or daos_size_t strictly.

File: utils/node_local_test.py
Comment (mjmac) [reply utils/node_local_test.py:1952]: I only changed this in the places where either a uuid or label would work, in order to prefer the label when it's present.

File: src/control/cmd/daos/util.h
Comment (mjmac) [reply src/control/cmd/daos/util.h:80]: Fixed.

File: src/utils/daos_autotest.c
Comment (ashleypittman) [inline src/utils/daos_autotest.c:68]: missed one?

File: src/control/go.mod
Comment (kjacque) [inline src/control/go.mod:10]: Oops

File: src/control/cmd/daos/util.h
Comment (kjacque) [inline src/control/cmd/daos/util.h:8]: I've only ever seen these logging macros defined in C files. It's probably OK here since this header has a very limited use, but I'll leave it to others with a better understanding of the logging system to comment.

Also noticed you didn't fence off the file with #ifndef ... #endif to avoid its contents being accidentally imported more than once. Again, may not be strictly required for this one, but still good practice.

File: src/control/cmd/daos/container.go
Comment (kjacque) [inline src/control/cmd/daos/container.go:189]: Should it be "container name or UUID" ?

File: src/control/cmd/daos/flags.go
Comment (kjacque) [inline src/control/cmd/daos/flags.go:35]: Any reason to not put this in the test file?

Comment (mjmac) [reply src/control/cmd/daos/flags.go:35]: Yes. You can't `import "C"` in test files. You can, however, test code that uses cgo, though. I don't understand the restriction, but it's well-known, e.g. https://github.com/golang/go/issues/4030

File: src/control/go.mod
Comment (mjmac) [reply src/control/go.mod:10]: Gah. How did that sneak back in?!?

File: src/control/cmd/daos/container.go
Comment (mjmac) [reply src/control/cmd/daos/container.go:189]: Yep. Will fix this and make sure it's consistent with the pool help strings.

File: src/control/cmd/daos/util.h
Comment (mjmac) [reply src/control/cmd/daos/util.h:8]: Will add the fencing. I think the theory behind adding it to .c files elsewhere is that a given header could be included from different files, but in this case as you say it's only going to be used in the daos tool. If we needed to enable different logging facilities from different parts of the tool this could be a problem, but I don't think that's going to happen. Mostly just trying to cut down on boilerplate.

File: src/control/cmd/daos/container.go
Comment (mjmac) [reply src/control/cmd/daos/container.go:189]: Oh. Actually, this is the create command, and this argument is for optionally supplying the container UUID to be used on create. There has been some discussion about removing this for the same reasons we removed it from the pool create flow, but for the moment we have to support it.

Comment (ashleypittman) [reply src/control/cmd/daos/container.go:189]: On this topic I've got another proposal to be able to pick up the pool from the path, so container create could work without a pool uuid/label set on the command line which will impact this code I suspect.

https://github.com/daos-stack/daos/pull/5969

Comment (mjmac) [reply src/control/cmd/daos/container.go:190]: Had a look at that PR. Will be easier to integrate that feature into the new tool after this one lands, as it cleans up a bunch of stuff.

Comment (ashleypittman) [reply src/control/cmd/daos/container.go:190]: I thought so, that's why I've paused it but I didn't want you to add something now that would make it more difficult later.

**Review by johannlombardi — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Pull request merged**
